### PR TITLE
[WIP] CUPS plugin

### LINKED ISSUE #857: CUPS information
CUPS provides a lot of interesting statistics (jobs in queue, printer statuses, printing frequency), and can benefit from alerts (for example, if print jobs start piling up or one gets stuck).

It would be very nice to see a CUPS plugin for netdata.

### PR DESCRIPTION
**Pull request by simonnagl**
This is a work in progress.

Closes #857

### CODE REVIEW

**Comment by ktsaou** [comment]
@simonnagl another way to implement this is to make it an external plugin. Actually we need to make this for nfacct too.

This will have a few pros and cons:

### pros

- fewer libraries to link netdata to
- in case an external library crashes, the server will survive
- in case a plugin/external library has a memory leak, the server will not be affected
- in case a plugin/external library corrupts memory, the server will not be affected

## cons

- it will be slower, since all metrics will have to be converted to text and parsed from text back (we could develop a binary plugin protocol though to avoid all this parsing - and even a shared memory protocol where data will not be copied from process to process)
- a lot of processes to in `top` and `ps`

However, the way to do it could be very simple. If you check your code, you use netdata functions like these:

```c
       RRDSET *st = rrdset_find_byname("cups.jobs");
        if (unlikely(!st))
        {
            st = rrdset_create("cups", "jobs", NULL, "jobs", NULL, "Total CUPS job number", "jobs",
                               3001, update_every, RRDSET_TYPE_LINE);

            rrddim_add(st, "jobs", "jobs", 1, 1, RRDDIM_ABSOLUTE);
        }
        else
            rrdset_next(st);

        rrddim_set(st, "jobs", (collected_number) num_jobs_total);
        rrdset_done(st);

        if (unlikely(netdata_exit))
            break;
    }
```

So, we could provide a set of these functions that will just `printf` the relative lines to `stdout`, a dummy `main()` and this can turned to an external plugin...

This is what I plan to do for nfacct...

**Comment by simonnagl** [comment]
You are right.

> fewer libraries to link netdata to
Of course `netdata` must not have cups as a hard dependency.
I knew that and just startet implementing because we do not support external C plugins yet.

**Comment by ktsaou** [comment]
nice commit!

**Comment by simonnagl** [comment]
> nice commit!

Thank you! I will notify if I think this is ready to be merged.

**Comment by simonnagl** [comment]
Hi @ktsaou, 

I have a Problem I am not able to resolve. I hope you can help. The current version of the plugin produces this output for my machine:

```
CHART cups.jobs '' TotalCUPSjobnumber jobs jobs cups line 3000 1
DIMENSION jobs '' absolute 1 1
BEGIN cups.jobs
SET jobs = 13
END
BEGIN cups.jobs
SET jobs = 13
END
...
```

The dashboard does not show the chart. In the logs I see netdata starts the plugin but it does not exit.
Can you tell me what might be wrong?

File: src/cups_plugin.c [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+    for(i = 1; i < argc ; i++) {
+        if(isdigit(*argv[i]) && !freq) {
+            int n = atoi(argv[i]);
+            if(n > 0 && freq < 86400) {
+                freq = n;
+                continue;
+            }
+        }
+        else if(strcmp("version", argv[i]) == 0 || strcmp("-version", argv[i]) == 0 || strcmp("--version", argv[i]) == 0 || strcmp("-v", argv[i]) == 0 || strcmp("-V", argv[i]) == 0) {
+            printf("freeipmi.plugin %s\n", VERSION);
```
Comment (ktsaou) [inline src/cups_plugin.c]: it says `freeipmi.plugin`

**Comment by ktsaou** [comment]
> The dashboard does not show the chart. In the logs I see netdata starts the plugin but it does not exit.

Are you sure it produces some output when it is started under netdata? If the plugin runs but does not produce any output, netdata will not stop it and will not complain.

I just checked your code. Please flush stdout. Most likely glibc is buffering the data to reduce the number of writes. So, just flush it between iterations.

**Comment by simonnagl** [comment]
> Please flush stdout

Thank you, this helped much. Now there are charts displayed. There are many errors in the code I know yet but I wanted it to get working first. I'll notify you to review when I fixed them.

**Comment by simonnagl** [comment]
Closed in favor of #5188

**Pull request closed without merging**